draft-irtf-cfrg-vdaf.md

```diff
@@ -1190,6 +1194,190 @@ how the nonces are chosen, but security requires that the nonces be unique. See
+               public_share: bytes,
+               host_input_share: bytes):
+    try:
+        host_prep_state = Vdaf.prep_init(
+            vdaf_verify_key,
+            0 if is_leader else 1,
+            agg_param,
+            nonce,
+            public_share,
+            host_input_share,
+        )
+        state CONTINUED(0, host_prep_state)
+    except:
+        state REJECTED
+    return None
+~~~
+
+If the state is `REJECTED`, then processing halts. Otherwise, if the state is
+`CONTINUE`, then processing continues using the following transition rule until
+a terminal state is reached.
+
+## Ping-Pong Topology (Only Two Aggregators)
+
+For VDAFs with precisely two Aggregators (i.e., `Vdaf.SHARES == 2`), the
+following "ping pong" communication pattern can be used. It is compatible with
+any request/response protocol, such as HTTP.
+
+Let us call the initiating party the "Leader" and the responding party the
+"Helper". The high-level idea is that the Leader and Helper will take turns
+running the computation locally until input from their peer is required:
+
+* For a 1-round VDAF (e.g., Prio3 ({{prio3}})), The Leader sends its prep share
+  to the Helper, who computes the prep message locally (the computations are
+  specified in the following sections), computes its output share, then sends
+  the prep message to the Leader. Preparation requires just one request from
+  the Leader to the Helper.
+
+* For a 2-round VDAF (e.g., Poplqr1 ({{poplar1}})), the Leader sends its
+  first-round prep share to the Helper, who replies with the first-round prep
+  message and its second-round prep share. In the next request, the Leader
+  computes its second-round prep share locally, computes its output share, and
+  sends the second-round prep message to the Helper. Finally, the Helper
+  computes its own output share. Preparation requires two requests.
+
+* In general, each request includes the Leader's prep share for the previous
+  round and/or the prep message for the current round; correspondingly, each
+  response consists of the prep message for the current round and the Helper's
+  prep share for the next round.
+
+The Aggregators proceed in this ping-ponging fashion until a step of the
+computation fails, indicating the report is invalid and should be rejected, or
+the last round is reached and each Aggregator computes an output share. All
+told there there are `ceil((Vdaf.ROUNDS+1)/2)` requests sent.
+
+~~~ transition
+State CONTINUED(round, host_prep_state)
+
+def ping_pong_prep_next(Vdaf,
+                        is_leader: bool,
+                        agg_param: Vdaf.AggParam,
+                        inbound_type: Optional[str],
+                        inbound: Optional[bytes]):
+    if inbound_type == "initialize":
+        prep_msg, peer_prep_share = None, inbound
+    elif inbound_type == "continue":
+        l = Vdaf.prep_size(agg_param, round)
+        prep_msg, prep_share = inbound[:l], inbound[l:]
```

Comment:
@branlwyd note that we're coalescing the previous prep message with the next prep share here. This is a change from https://github.com/ietf-wg-ppm/draft-ietf-ppm-dap/pull/461.

Reply:
This is tricky because we're trying to get as close to runnable code as possible. But your suggestion here matches @divergentdave's suggestion above pretty well, so let me see if I can finesse this.